### PR TITLE
Save/load

### PR DESCRIPTION
**Pull request by DierF**
Basic Save/Load feature using Serializer. And more test.

### CODE REVIEW

File: Serializer.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+        Parameters:
+        - diagram (Diagram): The diagram object containing entities and relations to be serialized.
+        - path (str): The file path where the JSON file will be saved.
+
+        Returns:
+        None
+
+        Raises:
+        - Any exceptions raised during JSON encoding or file writing.
```
Comment (timbmoser) [inline Serializer.py:23]: These need to be listed out.

Comment (timbmoser) [inline Serializer.py:23]: These need to be listed out.

File: Serializer.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+        Parameters:
+        - diagram (Diagram): The diagram object to populate with deserialized data.
+        - path (str): The file path of the JSON file to deserialize.
+
+        Returns:
+        None
+
+        Raises:
+        - Any exceptions raised during JSON decoding or file reading.
```
Comment (timbmoser) [inline Serializer.py:43]: These need to be listed out.

File: Serializer.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                        setattr(entity, property_name, property_val)
+                    diagram._entities[name] = entity
+            elif attr_name == 'relations':
+                for properties in attr_obj:
+                    relation = Relation()
+                    for property_name, property_val in properties.items():
+                        setattr(relation, property_name, property_val)
+                    diagram._relations.append(relation)
+
+    def __read(self, path: str) -> str:
```
Comment (timbmoser) [inline Serializer.py:61]: Needs doc string

File: Serializer.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        return open(path, 'r').read()
+    
+    def __write(self, path: str, content: str) -> None:
```
Comment (timbmoser) [inline Serializer.py:64]: Needs doc string.

File: Controller.py [hunk 5]
```diff
@@ -1,17 +1,27 @@
...
     def run(self) -> None:
         while not self._shouldQuit:
             s = self._input.readLine()
             if s == 'quit' or s == 'exit':
                 self._shouldQuit = True
-            self._output.write('Last cmd: {}'.format(s))
\ No newline at end of file
+            self._output.write('Last cmd: {}'.format(s))
+        
+    def save(self, path: str) -> None:
```
Comment (timbmoser) [inline Controller.py:21]: Needs Doc String.

File: Controller.py [hunk 6]
```diff
@@ -1,17 +1,27 @@
...
+        self._serializer.serialize(diagram=self._diagram, path=path)
+
+    def load(self, path: str) -> None:
```
Comment (timbmoser) [inline Controller.py:24]: Needs Doc String.

**Review by timbmoser — CHANGES_REQUESTED** [review]
Mostly just doc string things need updated/added.

Good catch on lines 46 and 47 of DiagramTest.py.

File: Serializer.py (hunk 4 above)
Comment (DierF) [reply Serializer.py:96]: Done

**Review by timbmoser — APPROVED** [review]

**Pull request merged**